### PR TITLE
Set the output size when slicing a dynamic dimension in tf2xla Slice

### LINKED ISSUE #110789: [XLA] Heap corruption (free(): invalid next size) when tf.slice produces zero-sized output from dynamically-shaped input
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf_nightly 2.21.0-dev20260112 (v1.12.1-135334-g20c4833e3b8)

### Custom code

Yes

### OS platform and distribution

Ubuntu 24.04.3 LTS

### Mobile device

_No response_

### Python version

3.11.14

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

12.9

### GPU model and memory

NVIDIA GeForce RTX 5090

### Current behavior?

When calling tf.slice with a zero-sized size argument on a dynamically-shaped tensor inside a jit_compile=True function:
- Eager mode succeeds and returns an empty tensor of shape (0, 1), consistent with expected TensorFlow semantics.
- TF-XLA mode (@tf.function(jit_compile=True)) crashes with a heap corruption error

tested with command: `CUDA_VISIBLE_DEVICES="" python3 test_code.py`

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf

class Model(tf.keras.Model):
    @tf.function(jit_compile=True)
    def call(self, inputs, training):
        # tf.where produces dynamic shape (None, 3)
        indices = tf.where(inputs)
        # top_k on dynamic shape -> (None, 2)
        values, _ = tf.math.top_k(indices, k=2, sorted=False)
        # slice with zero size on dynamic shape -> (0, 1)
        return tf.slice(values, tf.constant([0, 0]), tf.constant([0, 1]))

m = Model()
inp = np.arange(0, 20).reshape([2, 2, 5]).astype(np.int64)
y = m(inp, training=False)
```

### Relevant log output

```shell
free(): invalid next size (fast)
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by vishwakt**
Fixes #110789.

## Problem

Slicing a tensor that carries a dynamic dimension, for example the output of `tf.where`, with constant `begin`/`size` under `jit_compile=True` returns silently wrong results. Verified on TF 2.21.0 and tf-nightly 2.22.0-dev20260823 (XLA:CPU):

| computation | eager | XLA |
|---|---|---|
| `where` then `slice(size=[0, 1])` | (0, 1) | (0, 0) |
| `where` then `slice(size=[2, 1])` | (2, 1) | (0, 1) |
| `where`, `top_k`, `slice(size=[2, 1])` | (2, 1) | (0, 2) |

The nonzero-size case silently drops all the data. On Linux x86 the same defect manifests as heap corruption (`free(): invalid next size`, the crash in the issue), consistent with buffers whose dynamic size metadata disagrees with the requested slice.

The cause is in the constant `begin`/`size` path of the tf2xla `Slice` kernel: it emits a plain `xla::Slice` and leaves the output's dynamism to XLA's dynamic dimension inference. The HLO shows the slice output typed `s64[<=2,1]` from an `s64[<=20,3]` input, and the propagated runtime size for the partially sliced dynamic dimension does not implement the slice semantics. The existing dynamism handling in this path only covers dynamic values of the `size` operand, not dynamic dimensions of the input.

## Fix

After emitting the slice, set the dynamic size of the output explicitly for every dynamic input dimension: a dimension holding `dynamic_size` valid elements contributes `clamp(dynamic_size - begin, 0, size)` elements to the slice. A full slice of a dynamic dimension keeps its current behavior, since the clamp then evaluates to the input's dynamic size. This mirrors how the sibling kernels manage dynamism explicitly with `SetDimensionSize` (the non-constant-size branch of this same file already does), and follows the `GetShape`/`is_dynamic_dimension` pattern used by `split_op.cc`.

When `begin + size` exceeds the runtime dynamic size, the result is clamped rather than raising the error eager mode produces, matching XLA's established policy for runtime bounds under compilation, since compiled programs cannot raise data-dependent errors.

## Test

`testSliceOfDynamicDimension` in `compiler/tests/slice_ops_test.py` slices the output of `where` under the XLA test harness and checks a nonzero and a zero sized slice against the eager values and shapes (expected values verified against eager semantics). Both assertions fail against the current wheels, matching the mismatch table above.

### CODE REVIEW

**Review by adi-IL — COMMENTED** [review]
Tested locally on Linux x86_64.

When slicing a dynamic dimension under XLA (such as from a `tf.where` mask), the inferred dynamic slice length was unconstrained. Computing `output_size = Clamp(0, input_dynamic_size - begin, requested_size)` and setting the dimension size via `xla::SetDimensionSize` preserves the dynamic dimension contract. The regression test in `slice_ops_test.py` passes.

**Review by dmiltr3 — APPROVED** [review]

**Comment by nithyak0204** [comment]
Hi @vishwakt, Seems auto-merge is not happening but the changes are merged into master now, so we can close this. Thank you for the PR.
